core/reflection.jvm/resources/META-INF/proguard/kotlinreflect.pro

```diff
@@ -0,0 +1,14 @@
+##---------------Begin: Kotlin-reflect ----------
+# Keep Metadata annotations so they can be parsed at runtime.
+-keep class kotlin.Metadata { *; }
+
+# Keep kotlin-reflect internals.
+-keep class kotlin.reflect.jvm.** { *; }
```

Comment:
This seems way too broad. At least, we can try _public_ APIs only.

Reply:
Let me try it in our project. To be totally honest I'm not convinced this is truly needed, but it's frequently suggested as a fix in the community. The last comprehensive list we were using internally was based on this:

https://github.com/square/moshi/pull/883

Though all of those are public classes as well, so it would work. @udalov any advice on what scope should be kept?

Reply:
The list from the issue mentioned above is actually the combination of all services and implementations used in kotlin-reflect -- see `META-INF/services` in `kotlin-reflect.jar`. Shouldn't they be handled correctly in some other way by Proguard/R8? E.g. is it possible to "keep" the directory `META-INF/services`, and everything it references, to avoid enumerating classes used there?

Other than that, we do need to fully keep the contents of packages `kotlin.reflect.full` and `kotlin.reflect.jvm`, but not their subpackages (i.e. `kotlin.reflect.jvm.internal`).

Reply:
Types declared inside `META-INF/services/` are handled automatically provided that the code which calls into `ServiceLoader` with that service type is not being removed. You shouldn't need to do anything for them.

Reply:
They're handled automatically by R8, but not proguard iirc. Can we have rules that are only picked up by one or the other?

Reply:
Yes. See kotlinx.coroutines for an example that targets R8 1.6 and newer only.

Reply:
Done in 08616fb

Reply:
Also tweaked to just keep service-loaded APIs, only in proguard or R8 <1.6 c15aac4

Reply:
Thanks for taking care of this thoroughly! Those specific rules that aim for only interface and services look great!